File: Advance/7.Find the shortest path in a graph using Dijkstra’s Algorithm/Solution.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int findMinDistance(vector<int>& dist, vector<bool>& visited, int n) {
    int min = INT_MAX, minIndex = -1;
    for (int v = 0; v < n; v++) {
        if (!visited[v] && dist[v] <= min) {
            min = dist[v];
            minIndex = v;
        }
    }
    return minIndex;
}

void dijkstra(vector<vector<int>>& graph, int n, int start) {
    vector<int> dist(n, INT_MAX);
    vector<bool> visited(n, false);

    dist[start] = 0;

    for (int count = 0; count < n - 1; count++) {
        int u = findMinDistance(dist, visited, n);
        visited[u] = true;

        for (int v = 0; v < n; v++) {
            if (!visited[v] && graph[u][v] && dist[u] != INT_MAX && dist[u] + graph[u][v] < dist[v]) {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }

    cout << "Vertex\tDistance from Source" << endl;
    for (int i = 0; i < n; i++) {
        cout << i << "\t" << dist[i] << endl;
    }
}

int main() {
    int n, start;
    cout << "Enter the number of vertices: ";
    cin >> n;

    vector<vector<int>> graph(n, vector<int>(n));
    cout << "Enter the adjacency matrix:" << endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> graph[i][j];
        }
    }

    cout << "Enter the starting vertex: ";
    cin >> start;

    dijkstra(graph, n, start);
    return 0;
}

File: Advance/7.Find the shortest path in a graph using Dijkstra’s Algorithm/Solution.py
import sys

def find_min_distance(dist, visited, n):
    min_val = sys.maxsize
    min_index = -1
    for v in range(n):
        if not visited[v] and dist[v] <= min_val:
            min_val = dist[v]
            min_index = v
    return min_index

def dijkstra(graph, n, start):
    dist = [sys.maxsize] * n
    visited = [False] * n

    dist[start] = 0

    for _ in range(n - 1):
        u = find_min_distance(dist, visited, n)
        visited[u] = True

        for v in range(n):
            if not visited[v] and graph[u][v] and dist[u] != sys.maxsize and dist[u] + graph[u][v] < dist[v]:
                dist[v] = dist[u] + graph[u][v]

    print("Vertex\tDistance from Source")
    for i in range(n):
        print(f"{i}\t{dist[i]}")

if __name__ == "__main__":
    n = int(input("Enter the number of vertices: "))
    graph = []
    print("Enter the adjacency matrix:")
    for i in range(n):
        graph.append(list(map(int, input().split())))

    start = int(input("Enter the starting vertex: "))
    dijkstra(graph, n, start)
